drivers/qcommsimmodem/radio-settings.c

```diff
@@ -0,0 +1,287 @@
+
+		g_free(pps);
+	}
+}
+
+static void qcom_msim_set_rat_mode(struct ofono_radio_settings *rs,
+			enum ofono_radio_access_mode mode,
+			ofono_radio_settings_rat_mode_set_cb_t cb,
+			void *data)
+{
+	struct radio_data *rd = ofono_radio_settings_get_data(rs);
+	struct cb_data *cbd = cb_data_new(cb, data, rs);
+	struct parcel rilp;
+	int pref = PREF_NET_TYPE_GSM_WCDMA;
+
+	switch (mode) {
+	case OFONO_RADIO_ACCESS_MODE_ANY:
+		pref = PREF_NET_TYPE_LTE_GSM_WCDMA;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_GSM:
+		pref = PREF_NET_TYPE_GSM_ONLY;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_UMTS:
+		pref = PREF_NET_TYPE_GSM_WCDMA;
+		break;
+	case OFONO_RADIO_ACCESS_MODE_LTE:
+		pref = PREF_NET_TYPE_LTE_GSM_WCDMA;
+		break;
+	}
+
+	if (pref != PREF_NET_TYPE_GSM_ONLY && multisim_rs_amount > 1) {
+		struct qcom_msim_pending_pref_setting *pps = g_try_new0(
+				struct qcom_msim_pending_pref_setting, 1);
+		int i;
+
+		pps->rs = rs;
+		pps->pref = pref;
+		pps->cbd = cbd;
+		pps->pending_gsm_pref_remaining = 0;
+
+		for (i = 0; i < MULTISIM_RS_LAST; i++) {
+			struct radio_data *current_rd;
+			struct qcom_msim_set_2g_rat *set_2g_rat_data;
+
+			if (multisim_rs[i] == rs)
+				continue;
+
+			current_rd = ofono_radio_settings_get_data(
+							multisim_rs[i]);
+			set_2g_rat_data = g_try_new0(
+						struct qcom_msim_set_2g_rat, 1);
+			set_2g_rat_data->pps = pps;
+			set_2g_rat_data->rs = multisim_rs[i];
+
+			g_ril_request_set_preferred_network_type(
+						current_rd->ril,
+						PREF_NET_TYPE_GSM_ONLY, &rilp);
+
+			if (g_ril_send(current_rd->ril,
+					RIL_REQUEST_SET_PREFERRED_NETWORK_TYPE,
+					&rilp, qcom_msim_set_2g_rat_cb,
+					set_2g_rat_data, g_free) == 0) {
+				ofono_error("%s: unable to set rat mode",
+								__func__);
+				pps->cbd = NULL;
+				g_free(cbd);
+				g_free(set_2g_rat_data);
+				CALLBACK_WITH_FAILURE(cb, data);
+				break;
+			} else {
+				pps->pending_gsm_pref_remaining += 1;
+			}
+		}
+	} else {
+		g_ril_request_set_preferred_network_type(rd->ril, pref, &rilp);
+
+		if (g_ril_send(rd->ril, RIL_REQUEST_SET_PREFERRED_NETWORK_TYPE,
+					&rilp, qcom_msim_set_rat_cb, cbd,
+					g_free) == 0) {
+			ofono_error("%s: unable to set rat mode", __func__);
+			g_free(cbd);
+			CALLBACK_WITH_FAILURE(cb, data);
+		}
+	}
+}
+
+static ofono_bool_t query_modem_rats_cb(gpointer user_data)
+{
+	ofono_bool_t modem_rats[OFONO_RADIO_ACCESS_MODE_LAST] = { FALSE };
+	struct cb_data *cbd = user_data;
+	ofono_radio_settings_modem_rats_query_cb_t cb = cbd->cb;
+	struct ofono_radio_settings *rs = cbd->user;
+	struct radio_data *rd = ofono_radio_settings_get_data(rs);
+
+	modem_rats[OFONO_RADIO_ACCESS_MODE_GSM] = TRUE;
+	modem_rats[OFONO_RADIO_ACCESS_MODE_UMTS] = TRUE;
+
+	if (ofono_modem_get_boolean(rd->modem, MODEM_PROP_LTE_CAPABLE))
+		modem_rats[OFONO_RADIO_ACCESS_MODE_LTE] = TRUE;
+
+	CALLBACK_WITH_SUCCESS(cb, modem_rats, cbd->data);
+
+	g_free(cbd);
+
+	return FALSE;
+}
+
+static void qcom_msim_query_modem_rats(struct ofono_radio_settings *rs,
```

Comment:
This implementation is the same as the rilmodem one. Please declare ril_query_modem_rats in "drivers/rilmodem/radio-settings.h" and use it instead.